runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -3515,106 +3802,1263 @@ private void doKafkaData(
+                        .typeId(kafkaTypeId)
+                        .offsetCommit(o -> o
+                            .topic(kp.topic)
+                            .progress(p -> p
+                                .partitionId(kp.partitionId)
+                                .partitionOffset(metadata.sequence)
+                                .metadata(offsetMetadataToString(metadata)))
+                            .generationId(delegate.generationId)
+                            .leaderEpoch((int) delegate.leaderEpochs.get(offsetKey)))
+                        .build();
+
+                    doKafkaData(traceId, authorization, budgetId, DATA_FLAG_COMPLETE, initialOffsetCommit);
+                });
+            }
+            else if (wasOpen)
+            {
+                if (delegate.initialisablePartitions.isEmpty())
+                {
+                    final int packetId = unackedPacketIds.remove();
+                    if (delegate.partitions.containsKey(packetId))
+                    {
+                        delegate.partitions.remove(packetId);
+                    }
+                    else if (delegate.retainedPartitions.containsKey(packetId))
+                    {
+                        delegate.retainedPartitions.remove(packetId);
+                    }
+
+                    delegate.doMqttFlush(traceId, authorization, 0, 0, packetId);
+                }
+                else
+                {
+                    // Receiving initial offset commit acks
+                    delegate.initialisablePartitions.remove(0);
+                    if (delegate.initialisablePartitions.isEmpty())
+                    {
+                        Flyweight mqttBeginEx = mqttSessionBeginExRW.wrap(sessionExtBuffer, 0, sessionExtBuffer.capacity())
+                            .typeId(mqttTypeId)
+                            .session(sessionBuilder ->
+                            {
+                                sessionBuilder
+                                    .flags(delegate.sessionFlags)
+                                    .expiry((int) TimeUnit.MILLISECONDS.toSeconds(delegate.sessionExpiryMillis))
+                                    .subscribeQosMax(MQTT_KAFKA_MAX_QOS)
+                                    .capabilities(MQTT_KAFKA_CAPABILITIES)
+                                    .clientId(delegate.clientId);
+
+                                delegate.offsets.values().forEach(o ->
+                                    o.packetIds.forEach(p -> sessionBuilder.appendPacketIds(p.shortValue())));
+                            }).build();
+
+                        delegate.doMqttBegin(traceId, authorization, 0, mqttBeginEx);
+                        delegate.session = new KafkaSessionStateProxy(originId, routedId, delegate);
+                        delegate.session.doKafkaBeginIfNecessary(traceId, authorization, 0);
+                    }
+                }
+            }
```

Comment:
This section is quite complex. Let's try moving some of the implementation reaction to `delegate.onXX` methods to let it know what happened and then let it react.

Try also to determine what are the invariants here in both sides of the if condition, to help improve the abstraction instead of treating each side as an independent linear codepath.